### PR TITLE
[full-ci] Vite

### PR DESCRIPTION
**Pull request by dschmidt**
## Description
This makes oC Web build with [Vite](https://vitejs.dev/).

To test this you need to do `docker-compose up -d`.
OAuth2 Application in oC 10 and OIDC Client in oCIS are setup automatically now.

- `pnpm vite` or `pnpm vite:ocis` -> https://host.docker.internal:9201/
   -> currently uses a shitty workaround because of https://github.com/owncloud/ocis/issues/5108
  -> possibly needs more dirs added to `server.proxy` in `vite.config.ts`
- `pnpm vite:oc10` -> http://host.docker.internal:8081/

## Motivation and Context
rollup builds are rather slow and we spend a lot of time waiting for builds during development.
vite supports Hot Module Reloading which makes changes instant. If there are errors they are shown in the browser and we don't need to switch to the terminal to notice them.
Moreover it's the standard tool in the Vue world by now.
It uses rollup as a bundling backend in production mode and has compatible plugin api, so we need to worry less about standard rollup plugins (because they are abstracted away by vite) and can still use more exotic rollup plugins.

## Open tasks:

- [x] Fix asset loading
- [x] Use a shared index.html for development and production: https://github.com/vbenjs/vite-plugin-html
- [x] Implement `oidc-callback.html`
- [x] Fix `process.env.PACKAGE_VERSION` replacement
- [x] Figure out how to replace define/requirejs for external apps
     - [x] during development, kinda works
       * [x] fix external app loading
       * [x] make sure development and production apps can be loaded
     - [x] in production
     - try to load `.js` and `.cjs` files as commonjs modules, use dynamic import only for `.mjs`. Possibly worth accepting TypeScript variants of those extensions to ease development of external apps with vite
 - [x] wire up with docker setup
 - [x] Check which rollup plugins are still needed and remove others
 - [x] Port `scripts` in `package.json`
 - [x] Import ODS?
 - [x] Apply owncloud-sdk code mod? -> https://www.npmjs.com/package/vite-plugin-replace
 - [x] Remove explicit style imports in design-system: https://vitejs.dev/config/shared-options.html#css-preprocessoroptions
 - [x] Fix `TypeError: r.pbkdf2Sync is not a function` on file download (in oCIS?) -> https://github.com/sodatea/vite-plugin-node-stdlib-browser ?
 - [x] Fix preview app
 - [x] Log warning when `@extend` is being used in a component
 - [x] Fix self referencing in modules to be able to revert the currently applied workarounds  https://github.com/vitejs/vite/issues/9731
## Followup tasks:
 - [ ] Reenable cssCodeSplit with Vue 3 (or when vue2 plugin supports it)
 - [ ] Use vitest 
 - [ ] Make browserlist available as esbuild target, possibly with: https://github.com/evanw/esbuild/issues/324
 - [ ] Port external apps to vite
   * [ ] ocis settings
   * [ ] web-app-gpx-viewer 🤓 
   * [ ] ....
   * [ ] ensure vue, vuex and luxon are not embedded in external apps
 - [ ] Clean up config.json files (?)
   * [ ] Ship `config.json.dist` files for vite, so users can provide their own `config.json` without changes to files tracked in git
   * [ ] Check if we can get rid of some `config.json` files
   * [ ] Add a helper to initialize `config.json` files with `.dist` files if they are not present. Could be handled in style`vite.config.ts` potentially
 -  [ ] Fix production rebuilds require `docker-compose restart oc10`. The dist folder is completely emptied and so our bind mounts break, dist cannot be mounted completely because of the file structure and where the php files need to be put
   -> do not merge dist folder with app.js and php files but have it in a separate sub folder
 - [ ] Add gzipped assets to production builds
 - [ ] enable history mode for `pnpm vite` / `pnpm vite:ocis` (not for oc10!)

### CODE REVIEW

**Comment by fschade** [comment]
Just read it on my cell… ❤️ it… I give it a try tomorrow! Thank you 😘

File: packages/web-app-files/src/services/folder.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ import {
   FolderLoaderSharedWithMe,
   FolderLoaderSharedWithOthers,
   FolderLoaderTrashbin
-} from './folder/'
+} from './folder/index'
```
Comment (kulmann) [inline packages/web-app-files/src/services/folder.ts:15]: Huh? So the appended `/` doesn't help to differentiate  folder `folder` and file `folder.ts`?

File: packages/web-runtime/src/layouts/Loading.vue [hunk 2]
```diff
@@ -8,7 +8,7 @@
     <oc-spinner size="xlarge" :aria-label="$gettext('Loading')" />
   </div>
 </template>
-<script>
+<script lang="ts">
 import { mapGetters } from 'vuex'
 
 export default {
```
Comment (kulmann) [inline packages/web-runtime/src/layouts/Loading.vue:14]: You forgot to wrap the object into `defineComponent()` ;-) At least that's what we usually do when we set `lang="ts"` on a script tag.

File: config/vite_oc10/config.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    "clientSecret": "VsrTbbeTPJ56e93eKpCdb6Wf5IGHD2meadlsDT1M9EpS3k7Y1ywTYgOhTkKZ0QTL",
+    "url": "http://host.docker.internal:8080/index.php/apps/oauth2/api/v1/token",
+    "authUrl": "http://host.docker.internal:8080/index.php/apps/oauth2/authorize",
+    "logoutUrl": "http://host.docker.internal:8080/index.php/logout"
+  },
+  "apps": [
+    "files",
+    "text-editor",
+    "pdf-viewer",
+    "search"
```
Comment (kulmann) [inline config/vite_oc10/config.json:15]: `preview` app still doesn't work right?

File: packages/web-runtime/src/layouts/Loading.vue (hunk 2 above)
Comment (dschmidt) [reply packages/web-runtime/src/layouts/Loading.vue:14]: no idea, why I added it at all - probably played around with something and only partially reverted it

File: packages/web-app-files/src/services/folder.ts (hunk 1 above)
Comment (dschmidt) [reply packages/web-app-files/src/services/folder.ts:15]: I'm not sure why I had to add this - needs double checking

File: config/vite_oc10/config.json (hunk 3 above)
Comment (dschmidt) [reply config/vite_oc10/config.json:15]: ack, some import/export issue somewhere afaict, just had no time to look into it so far

**Comment by ownclouders** [comment]
Results for <strong>e2e-tests oC10</strong> https://drone.owncloud.com/owncloud/web/30397/11/1
<details><summary>:boom: To see the trace, please open the link in the console ...</summary>

<p>


#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/30397/tracing/alice-can-share-this-weeks-meal-plan-with-all-parents-alice-2022-12-5-09-52-16.zip 

#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/30397/tracing/file-alice-2022-12-5-09-57-40.zip 

#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/30397/tracing/folder-alice-2022-12-5-09-56-35.zip 

#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/30397/tracing/public-link-alice-2022-12-5-09-53-18.zip 

#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/30397/tracing/re-sharing-alice-2022-12-5-09-54-25.zip 

#### npx playwright show-trace https://cache.owncloud.com/public/owncloud/web/30397/tracing/search-in-personal-spaces-brian-2022-12-5-09-55-30.zip 


</p></details>
:boom: The e2e-oc10 tests pipeline failed. The build has been cancelled.

File: packages/design-system/src/styles/styles2.scss [hunk 4]
```diff
@@ -0,0 +1 @@
+@import "styles"
```
Comment (dschmidt) [inline packages/design-system/src/styles/styles2.scss:1]: Give file a better name, maybe `stylesWrapper.scss`  and explain it's needed so Vite does not complain the file is already loaded.

File: vite.config.ts [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+              map: null
+            }
+          }
+        }
+      },
+      {
+        name: 'ejs-pre',
+        transformIndexHtml: {
+          enforce: "pre",
+          transform(html, ...args) {
```
Comment (dschmidt) [inline vite.config.ts:176]: remove ...args

File: config/vite_ocis/config.json [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    "scope": "openid profile email"
+  },
+  "apps": [
+    "files",
+    "text-editor",
+    "pdf-viewer",
+    "search",
+    "external",
+    "user-management",
+    "preview"
```
[the remark this answers is not in the document]
Comment (kulmann) [reply config/vite_ocis/config.json:18]: sorry... not in the file I mentioned. Anyway, please add the preview as follows to the external apps and remove it from the apps array:
```json
{
      "id": "preview",
      "path": "web-app-preview",
      "config": {
        "mimeTypes": ["image/tiff","image/bmp","image/x-ms-bmp"]
      }
    }
```

**Review by kulmann — APPROVED** [review]
I think this still has some rough edges for devs:
- router is in hash mode (that's controlled server side at the moment, directly in ocis via modification of the index.html)
- public links are announced by the server and as such are not with port `9201` when running in `pnpm vite`-dev-mode.

Both of those are not blockers for merging the PR. But it's really good that running the dev setup with `vite` is not documented by this PR, yet. Since the production `build` and `build:w` still work the same in the dev setup, and that is what we've documented, I see no blockers for this PR.

Love how you modernized a whole lot of things. And HMR of course. 😁 ❤️ Let's get this merged asap.

**Comment by kulmann** [comment]
@diocas @elizavetaRa do you still rebase to `master`? If yes then you might experience some issues with your other apps - please reach out, we'll help you fix issues. Also, please note that `data.config.cdn` is now `data.buildConfig.cnd` in the `index.html` file. Again, if you need help fixing stuff, please reach out.

**Comment by pascalwengerter** [comment]
Lovely! oC web, welcome to the wild 20s of the 21 century :D I'm {proud of,happy for} y'all

**Pull request merged**